graphbolt/src/cuda/unique_and_compact_impl.cu

```diff
@@ -41,139 +47,481 @@ struct EqualityFunc {
+          unique_ids_offsets_ptr[i] = offsets_ptr[2 * i];
         }
-        auto new_dst_ids = torch::empty_like(dst_ids);
         THRUST_CALL(
-            gather, new_dst_ids_loc.get(),
-            new_dst_ids_loc.get() + dst_ids.size(0),
-            new_ids.data_ptr<int64_t>(), new_dst_ids.data_ptr<scalar_t>());
-        return std::make_tuple(real_order, new_src_ids, new_dst_ids);
+            gather, unique_ids_offsets_ptr,
+            unique_ids_offsets_ptr + unique_ids_offsets.size(0),
+            valid.data_ptr<int64_t>(), unique_ids_offsets_ptr);
+        at::cuda::CUDAEvent unique_ids_offsets_event;
+        unique_ids_offsets_event.record();
+        auto unique_ids =
+            torch::empty(offsets_ptr[2 * num_batches], src_ids[0].options());
+        CUDA_KERNEL_CALL(
+            _GetInsertedBatched, grid, block, 0, offsets_ptr[2 * num_batches],
+            indexes.data_ptr<int32_t>(), pointers_dev_ptr, offsets_dev_ptr,
+            map.ref(cuco::find), valid.data_ptr<int64_t>(),
+            unique_ids.data_ptr<index_t>());
+        auto mapped_ids =
+            torch::empty(offsets_ptr[3 * num_batches], unique_ids.options());
+        CUDA_KERNEL_CALL(
+            _MapIdsBatched, grid, block, 0, num_batches,
+            offsets_ptr[3 * num_batches], indexes.data_ptr<int32_t>(),
+            pointers_dev_ptr, offsets_dev_ptr, map.ref(cuco::find),
+            mapped_ids.data_ptr<index_t>());
+        std::vector<std::tuple<torch::Tensor, torch::Tensor, torch::Tensor>>
+            results;
+        unique_ids_offsets_event.synchronize();
+        for (int64_t i = 0; i < num_batches; i++) {
+          results.emplace_back(
+              unique_ids.slice(
+                  0, unique_ids_offsets_ptr[i], unique_ids_offsets_ptr[i + 1]),
+              mapped_ids.slice(
+                  0, offsets_ptr[2 * i + 1], offsets_ptr[2 * i + 2]),
+              mapped_ids.slice(
+                  0, offsets_ptr[2 * num_batches + i],
+                  offsets_ptr[2 * num_batches + i + 1]));
+        }
+        return results;
       }));
 }
 
+std::vector<std::tuple<torch::Tensor, torch::Tensor, torch::Tensor>>
+UniqueAndCompactBatched(
+    const std::vector<torch::Tensor>& src_ids,
+    const std::vector<torch::Tensor>& dst_ids,
+    const std::vector<torch::Tensor> unique_dst_ids, int num_bits) {
+  auto dev_id = cuda::GetCurrentStream().device_index();
+  static std::mutex mtx;
+  static std::unordered_map<decltype(dev_id), int> compute_capability_cache;
+  const auto compute_capability_major = [&] {
+    std::lock_guard lock(mtx);
+    auto it = compute_capability_cache.find(dev_id);
+    if (it != compute_capability_cache.end()) {
+      return it->second;
+    } else {
+      int major;
+      C10_CUDA_DRIVER_CHECK(cuDeviceGetAttribute(
+          &major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR, dev_id));
+      return compute_capability_cache[dev_id] = major;
+    }
+  }();
+  if (compute_capability_major >= 7) {
```

Comment:
@TristonC @chang-l Does this dispatch mechanism look right?

Reply:
I don't know how to structure the code so that I can make the CI passes. cuda atomic header can not be included when compiling for computa capability < 70.

```
In file included from /home/ubuntu/jenkins/workspace/dgl_PR-7239/third_party/cccl/libcudacxx/include/cuda/std/detail/libcxx/include/atomic:733,
                 from /home/ubuntu/jenkins/workspace/dgl_PR-7239/third_party/cccl/libcudacxx/include/cuda/std/atomic:18,
                 from /home/ubuntu/jenkins/workspace/dgl_PR-7239/graphbolt/../third_party/cccl/libcudacxx/include/cuda/atomic:14,
                 from /home/ubuntu/jenkins/workspace/dgl_PR-7239/graphbolt/../third_party/cuco/include/cuco/detail/open_addressing/kernels.cuh:22,
                 from /home/ubuntu/jenkins/workspace/dgl_PR-7239/graphbolt/../third_party/cuco/include/cuco/detail/open_addressing/open_addressing_impl.cuh:21,
                 from /home/ubuntu/jenkins/workspace/dgl_PR-7239/graphbolt/../third_party/cuco/include/cuco/static_map.cuh:21,
                 from /home/ubuntu/jenkins/workspace/dgl_PR-7239/graphbolt/src/cuda/unique_and_compact_impl.cu:14:
/home/ubuntu/jenkins/workspace/dgl_PR-7239/third_party/cccl/libcudacxx/include/cuda/std/detail/libcxx/include/support/atomic/atomic_cuda.h:12:4: error: #error "CUDA atomics are only supported for sm_60 and up on *nix and sm_70 and up on Windows."
   12 | #  error "CUDA atomics are only supported for sm_60 and up on *nix and sm_70 and up on Windows."
      |    ^~~~~
```